gobblin-dataset-config/src/main/java/gobblin/dataset/config/FileBasedConfigStore.java

```diff
@@ -0,0 +1,142 @@
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.dataset.config;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.log4j.Logger;
+
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+
+/**
+ * This class is the implementation for the filed based config store. 
+ * @author mitu
+ *
+ */
+public class FileBasedConfigStore implements ConfigStore {
+
+  private static final Logger LOG = Logger.getLogger(FileBasedConfigStore.class);
+  public static final String CONF_FILE = "application.conf";
+  private final String configLocation;
+  private final String scheme;
+
+  private String latestVersion;
+  private String loadedConfigVersion;
+  private RawConfigMapping rawConfigs;
+  private boolean initialized = false;
+
+  /**
+   * @param location configuration store location
+   * @param scheme the scheme name, example DAI-ETL, DALI, Espresso
+   */
+  public FileBasedConfigStore(String location, String scheme) {
+    this.configLocation = location;
+    this.scheme = scheme;
+
+    File[] subfiles = new File(this.configLocation).listFiles();
+    List<String> subdirs = new ArrayList<String>();
+    for(File f: subfiles){
+      if(f.isDirectory()){
+        subdirs.add(f.getName());
+      }
+    }
+    
+    latestVersion = VersionPattern.getLatestVersion(subdirs);
+    if(latestVersion==null){
+      throw new RuntimeException("Can not find any valid version in " + this.configLocation);
+    }
+  }
+  
+  public FileBasedConfigStore(File file, String scheme) {
+    this(file.getAbsolutePath(), scheme);
+  }
+
+  @Override
+  public String getLatestVersion() {
+    return this.latestVersion;
+  }
+
+  @Override
+  public String getScheme() {
+    return this.scheme;
+  }
+
+  @Override
+  public synchronized void loadConfigs() {
+    loadConfigs(this.getLatestVersion());
+  }
+
+  @Override
+  public synchronized void loadConfigs(String version) {
+    String configurationFile = this.configLocation + "/" + version + "/" + CONF_FILE;
+    LOG.info("Trying to load configuration file: " + configurationFile);
+    File configToLoad = new File(configurationFile);
+    if(!configToLoad.isFile() || !configToLoad.canRead()){
+      throw new IllegalArgumentException("File does not exists : " + configToLoad.getAbsolutePath());
+    }
+    Config config = ConfigFactory.parseFile(configToLoad).resolve();
+    rawConfigs = new RawConfigMapping(config);
+    this.loadedConfigVersion = version;
+    initialized = true;
+  }
+
+  private void initialCheck(String version) {
+    if (!this.initialized) {
+      this.loadConfigs(version);
+    }
+
+    if (!this.loadedConfigVersion.equals(version)) {
```

Comment:
It is the ConfigAccessor responsibility, not the ConfigStore responsibility to ensure using a consistent version. The ConfigStore should expose only the low-level primitives to access the config.